src/blockchain/Blockchain.ts

```diff
@@ -141,28 +211,158 @@ export default class Blockchain implements IBlockchainApi {
+    const options = parseSubscriptionOptions(opts)
+    return submitSignedTx(tx, options)
+      .catch(async (reason: Error) => {
+        if (reason.message === 'Recoverable') {
+          return submitSignedTx(await this.reSignTx(identity, tx), options)
+        }
+        throw reason
+      })
+      .catch(async (reason: Error) => {
+        if (reason.message === 'Recoverable') {
+          return submitSignedTx(await this.reSignTx(identity, tx), options)
+        }
+        throw reason
+      })
+  }
+
+  /**
+   * [ASYNC] Signs and submits the SubmittableExtrinsic with optional resolving and rejection criteria.
+   *
+   * @param identity The [[Identity]] that we retrieve the nonce for.
+   * @param tx The generated unsigned SubmittableExtrinsic to submit.
+   * @param opts Optional [[SubscriptionPromiseOptions]].
+   * @returns Promise Result of The Extrinsic.
+   *
+   */
   public async submitTx(
     identity: Identity,
     tx: SubmittableExtrinsic,
-    opts: SubscriptionPromiseOptions = {}
+    opts?: SubscriptionPromiseOptions
   ): Promise<SubmittableResult> {
     const signedTx = await this.signTx(identity, tx)
-    return Blockchain.submitSignedTx(signedTx, opts)
+    return this.submitSignedTx(identity, signedTx, opts)
+  }
+
+  /**
+   * [ASYNC] Retrieves the Nonce for Transaction signing for the specified account, and increments the in accountNonces mapped Index.
+   *
+   * @param accountAddress The address of the identity that we retrieve the nonce for.
+   * @param reset Specify whether the entry for the account is outdated and has to be reset.
+   * @returns [[Index]] representation of the Tx nonce for the identity.
+   *
+   */
+  public async getNonce(accountAddress: string, reset = false): Promise<Index> {
+    if (reset) {
+      this.accountNonces.delete(accountAddress)
+    }
+    const initialQuery = this.accountNonces.get(accountAddress)
+    if (initialQuery) {
+      this.accountNonces.set(
+        accountAddress,
+        this.api.registry.createType('Index', initialQuery.addn(1))
+      )
+      return initialQuery
+    }
+    const chainNonce = await this.api.rpc.system
+      .accountNextIndex(accountAddress)
+      .catch((reason) => {
+        log.error(
+          `Onchain Nonce Retrieval failed for account ${accountAddress}\nwith Reason: ${reason}`
+        )
+        throw Error(`Chain failed to retrieve nonce for : ${accountAddress}`)
+      })
+    const secondQuery = this.accountNonces.get(accountAddress)
+    if (chainNonce && (!secondQuery || secondQuery.lte(chainNonce))) {
+      this.accountNonces.set(
+        accountAddress,
+        this.api.registry.createType('Index', chainNonce.addn(1))
+      )
+      return chainNonce
+    }
+    if (secondQuery) {
+      this.accountNonces.set(
+        accountAddress,
+        this.api.registry.createType('Index', secondQuery.addn(1))
+      )
+      return secondQuery
+    }
+    throw Error(`Nonce Retrieval Failed for : ${accountAddress}`)
```

Comment:
So in pseudo code this does:

```
nonce := get nonce from store
if nonce was not in store:
    chain_nonce := get nonce from chain
    stored_nonce := get nonce from store
    nonce = max(chain_nonce, stored_nonce)
store nonce + 1
return nonce
```

Might be easier to read since it only returns once and it's written in pseudo code. 😉

Reply:
Yes, love it. I ditched the whole Index notation, we work with BN only now.